api/src/main/java/com/netflix/iceberg/encryption/EncryptionKeyMetadata.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.netflix.iceberg.encryption;
+
+import com.netflix.iceberg.Schema;
+import com.netflix.iceberg.types.Types;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Metadata stored in Iceberg that points to an encryption key in some key store implementation.
+ * <p>
+ * Note that under any circumstances, the encryption key used to decrypt the file itself
+ * should NEVER be stored in this blob.
+ */
+public interface EncryptionKeyMetadata {
+
+  Schema SCHEMA = new Schema(
+      Types.NestedField.required(10000, "key_metadata", Types.BinaryType.get()),
+      Types.NestedField.required(10001, "cipher_algorithm", Types.StringType.get()),
+      Types.NestedField.required(10002, "key_algorithm", Types.StringType.get()));
+
+  static Schema schema() {
+    return SCHEMA;
+  }
+
+  /**
+   * Header description of the key.
+   */
+  ByteBuffer keyMetadata();
+
+  /**
+   * Cipher algorithm used to decrypt the file with this file's encryption key.
+   */
+  String cipherAlgorithm();
+
+  /**
+   * Algorithm to use when converting the key bytes into key material that can be used for
+   * creating cryptographic ciphers.
+   */
+  String keyAlgorithm();
```

Comment:
So we ended up adding the key algorithm here too. The reasoning for this is not obvious until we actually try doing the encryption itself in https://github.com/mccheah/incubator-iceberg/pull/1.

In a previous version of the proposal, we had a `FileEncryptionMetadata` field in the `DataFile` schema. The `FileEncryptionMetadata` contained an `EncryptionKeyMetadata`, which was just the byte header, and a separate field for the `cipherAlgorithm` used to generate cryptographic ciphers. But after using the `cipherAlgorithm` to generate the `keyMaterial` (see [here](https://github.com/mccheah/incubator-iceberg/pull/1/files#diff-d2da4db4af7249e243edbc9034f0cb68R100)), previously we didn't provide the `cipherAlgorithm` in `PhysicalEncryptionKey` because `cipherAlgorithm` was taken to be in `FileEncryptionMetadata`. In fact, in the old proposal the implementation of `KeyManager` had no way to share how it picks the `cipherAlgorithm` that would be associated with each `DataFile` object and their associated `FileEncryptionMetadata`s.

It is a bit difficult to explain, but the bottom line is that it's easier to keep everything consistent by having the `EncryptionKeyMetadata` embed both the `keyAlgorithm` and the `cipherAlgorithm`. The embedding can be done one of two ways:
1) Put them in the opaque bytes of the metadata and leave it up to the `keyManager` to return them in `PhysicalEncryptionKey` via decoding accordingly.
2) Make these concepts core parts of the `EncryptionKeyMetadata` schema.

The advantage of (2) is really that one can expect every encoding of key metadata to have to include the cipher and key algorithms. In other words, cipher and key algorithms for each key are _universal_ metadata tags. So there's no point forcing every key manager implementation to either do the same metadata decoding, or else have them need to look up these algorithms in their own backing storage.

The `cipherAlgorithm` and `keyAlgorithm` tags are more or less non-sensitive; we don't leak any security by exposing the knowledge that files were encrypted with a certain algorithm, or that multiple files were encrypted using the same algorithm. Thus while we're trying to be minimalist with regards to the amount of metadata we're trying to add to `DataFile`s, one can argue that these extra fields make sense to be included in the schema.

Reply:
I agree with your description of the choice, but I don't think I agree with the conclusion. While key and cipher algorithms are universal, how an encryption scheme chooses to manage them is not.

A great example that demonstrates what I'm talking about is the PR you linked to. The cipher algorithm is [configured by a table property](https://github.com/mccheah/incubator-iceberg/pull/1/files#diff-d2da4db4af7249e243edbc9034f0cb68R75), and the key algorithm is [determined by the cipher algorithm](https://github.com/mccheah/incubator-iceberg/pull/1/files#diff-d2da4db4af7249e243edbc9034f0cb68R99). So I can easily imagine a scheme where we don't store a string for each data file with its cipher and key algorithm, but instead initialize the key manager from table properties. Another reasonable scheme is to use an algorithm based on the encryption classes that are loaded.

Given that there are reasonable options that don't require adding 2 strings per file and that those strings add overhead, I'd prefer to go with option 1. That really simplifies the metadata that Iceberg has to store and adds little overhead.

Reply:
I don't think it is. This assumes that those are going to be strings, but they could easily be 1-byte algorithms IDs known to the crypto provider. So trying to standardize on String can be wasteful. It would be easy to use a combination of ids and table properties to use a String, too. `encryption.key-algorithms.1=...` and `encryption.cipher-algorithms.1=...` allow storing strings, but using IDs.

Any of these schemes is reasonable, so I'm reluctant to add per-file metadata.